src/nemos/solvers/_svrg_defaults.py

```diff
@@ -0,0 +1,548 @@
+    It is efficient for situations where X can fit in memory.
+    If batch_size is provided, the computation will be done by slicing the array.
+
+    Parameters
+    ----------
+    X :
+        Input data matrix (N x d).
+    y :
+        Output data vector (N,).
+    v :
+        Vector to be multiplied (d,).
+
+    Returns
+    -------
+    :
+        Result of the multiplication (X.T @ D @ X) @ v.
+    """
+    N, K = X.shape
+    out = jnp.zeros((K,))
+    for i in range(0, N, batch_size):
+        xb, yb = X[i : i + batch_size], y[i : i + batch_size]
+        out = out + xb.T.dot((0.17 * yb + 0.25) * xb.dot(v))
+    out = out / N
+    return out
+
+
+def _glm_softplus_poisson_l_smooth_with_power_iteration(
+    X: NDArray, y: NDArray, n_power_iters: int = 20, batch_size: Optional[int] = None
+):
+    """
+    Compute the largest eigenvalue of X.T @ D @ X using the power method.
+
+    Instead of calculating the full matrix and its eigenvalue directly, this function
+    uses the power iteration method, which is more memory efficient and scales better
+    with large datasets.
+
+    Parameters
+    ----------
+    X :
+        Input data matrix (N x d).
+    y :
+        Output data vector (N,).
+    n_power_iters :
+        Maximum number of power iterations to use.
+    batch_size :
+        The batch size, if user provides one.
+
+    Returns
+    -------
+    :
+        The largest eigenvalue of X.T @ D @ X.
+    """
+
+    if batch_size is None:
+        batch_size = X.shape[0]
+
+    _, d = X.shape
+
+    # Initialize a random d-dimensional vector for power iteration
+    v = jnp.ones((d,))
+
+    # Run power iteration to approximate the largest eigenvalue
+    for _ in range(n_power_iters):
+        v_prev = v.copy()
+        v = _glm_softplus_poisson_l_smooth_multiply(X, y, v, batch_size)
+        v /= v.max()
+
+        # Check for convergence
+        if jnp.allclose(v_prev, v):
+            break
+
+    # Final eigenvalue calculation
+    v /= jnp.linalg.norm(v)
+    return _glm_softplus_poisson_l_smooth_multiply(X, y, v, batch_size).dot(v)
+
+
+def _glm_softplus_poisson_l_smooth(
+    X: NDArray, y: NDArray, batch_size: int, n_power_iters: Optional[int] = None
+) -> jnp.ndarray:
+    """
+    Calculate the smoothness constant `L` for a Poisson GLM with softplus inverse link.
+
+    Depending on whether `n_power_iters` is provided, this function either computes
+    the largest eigenvalue directly or uses the power method.
+
+    Parameters
+    ----------
+    X :
+        Input data matrix (N x d).
+    y :
+        Output data vector (N,).
+    batch_size :
+        The batch size, if user provides one.
+    n_power_iters :
+        Number of power iterations to use when finding the largest eigenvalue. If None,
+        the eigenvalue is calculated directly.
+
+    Returns
+    -------
+    :
+        Smoothness constant `L`.
+    """
+    if n_power_iters is None:
```

Comment:
do we want to raise a warning (or an exception) if it looks like this a bad idea because the matrices are too large?

Reply:
what is the threshold for this? If the matrix is too big the user will get a runtime error, we can try catch that, and raise a more understandable RuntimeError, saying: "Computing the optimal stepsize resulted in a RuntimeError. This is likely because the model design matrix is too large...." and add a possible solution.